### PR TITLE
8318082: ConcurrentModificationException from IndexWriter

### PR DESCRIPTION
**Pull request by shipilev**
Unclean backport to fix a customer-visible bug in Javadoc tool. The uncleanliness comes from the missing refactoring ([JDK-8309566](https://bugs.openjdk.org/browse/JDK-8309566), that looks unnecessary to backport.

Maybe @jonathan-gibbons and @pavelrappo want to take a look.

Additional testing:
 - [x] New test fails without the product fix, passes with it
 - [x] `jdk/javadoc` still passes

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8318082](https://bugs.openjdk.org/browse/JDK-8318082) needs maintainer approval

### Issue
 * [JDK-8318082](https://bugs.openjdk.org/browse/JDK-8318082): ConcurrentModificationException from IndexWriter (**Bug** - P4 - Approved)

### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - Committer)
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21u-dev.git pull/23/head:pull/23` \
`$ git checkout pull/23`

Update a local copy of the PR: \
`$ git checkout pull/23` \
`$ git pull https://git.openjdk.org/jdk21u-dev.git pull/23/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23`

View PR using the GUI difftool: \
`$ git pr show -t 23`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21u-dev/pull/23.diff">https://git.openjdk.org/jdk21u-dev/pull/23.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk21u-dev/pull/23#issuecomment-1854493083)

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 1]
```diff
@@ -399,6 +399,20 @@ public TagletWriterImpl getTagletWriterInstance(TagletWriterImpl.Context context
         return new TagletWriterImpl(this, context);
     }
 
+    /**
+     * {@return true if the page written by this writer should be indexed,
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:403]: Use `{@code true}`

Comment (shipilev) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:403]: Thanks, but this is a backport, which I would like to keep this in sync with mainline: https://github.com/openjdk/jdk/blob/cf948548c390c42ca63525d41a9d63ff31349c3a/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java#L415

I wonder if we want to change this in mainline and backport separately?

**Review by pavelrappo — APPROVED** [review]
Output from `git range-diff pull/23^! fc29a2e^!` and this backport look fine.

**Comment by shipilev** [comment]
Still looking for formal (R)eviewers here, then I could apply for maintainer approval.

**Comment by shipilev** [comment]
@phohensee, @jerboaa, maybe?

**Review by simonis — APPROVED** [review]
Looks good.

**Comment by shipilev** [comment]
/approval request Unclean backport to fix a customer-visible bug in Javadoc tool. There are contextual differences that prevent it from being clean. 21u PR acked by Jonathan, Pavel, and Volker.

**Comment by shipilev** [comment]
Thanks all!

/integrate

**Pull request closed without merging**